### ISSUE TITLE
Middleware not protecting application

### ISSUE THREAD

**Issue by IncognitoTGT** [nextauthjs/next-auth#10320]
### Environment

```
System:
    OS: macOS 14.1
    CPU: (8) arm64 Apple M2
    Memory: 37.39 MB / 24.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 21.6.2 - /opt/homebrew/bin/node
    npm: 10.2.4 - /opt/homebrew/bin/npm
    pnpm: 8.15.4 - /opt/homebrew/bin/pnpm
  Browsers:
    Chrome: 122.0.6261.129
    Chrome Canary: 124.0.6357.0
    Safari: 17.1
    Safari Technology Preview: 17.4
  npmPackages:
    @auth/core: ^0.28.0 => 0.28.0
    @auth/prisma-adapter: ^1.5.0 => 1.5.0
    next: 14.2.0-canary.1 => 14.2.0-canary.1
    next-auth: 5.0.0-beta.15 => 5.0.0-beta.15
    react: ^18.2.0 => 18.2.0
```

### Reproduction URL

https://github.com/any-VM/anycode/tree/81aa6061c6331566cd235c4566fe92f31ad26972

### Describe the issue

I have auth.js setup with a database, with JWT being used for sessions, so that I can use middleware to authenticate. The application uses Keycloak auth. Sessions work, but the middleware does NOT protect the pages, and you can access the protected portions of the site.

Reopening of #10318 since it doesn't support gitlab urls

For it to work with middleware, I copied the guide from [here](https://authjs.dev/guides/upgrade-to-v5#edge-compatibility)
middleware.ts
```ts
import authConfig from "@/app/api/auth/[...nextauth]/auth.config";
import NextAuth from "next-auth";
export const { auth: middleware } = NextAuth(authConfig);
```
auth.config.ts
```ts
import Keycloak from "@auth/core/providers/keycloak";
import type { NextAuthConfig } from "next-auth";
const config: NextAuthConfig = {
  pages: {
    signIn: "/auth/login",
    verifyRequest: "/auth/verify",
    signOut: "/auth/logout",
    error: "/auth/error",
  },
  providers: [
    Keycloak({
      clientId: process.env.KEYCLOAK_ID,
      clientSecret: process.env.KEYCLOAK_SECRET,
      issuer: process.env.KEYCLOAK_ISSUER,
    }),
  ],
};
export default config;
```
route.ts
```ts
import NextAuth from "next-auth";
import { PrismaAdapter } from "@auth/prisma-adapter";
import prisma from "@/lib/prisma";
import authConfig from "./auth.config";
export const {
  handlers: { GET, POST },
  auth,
} = NextAuth({
  adapter: PrismaAdapter(prisma),
  session: { strategy: "jwt" },
  ...authConfig,
});
```

### How to reproduce

Clone the repository, fill in the env variables, and run the dev server. Go to /auth/login, and when you do, it will work and you'll see your email in the corner of the dashboard page. BUT, when you log out, you'll see that you can still access the dashboard, and in the corner, it just doesn't show anything.

### Expected behavior

The middleware should be requiring auth and not allowing public access to the dashboard.

**Comment by IncognitoTGT** [comment]
Update: It seems like middleware seems to be the bug here, since I tried it the normal way, and i'm still having the same problem.

tldr: Auth.js Middleware is broken

**Comment by IncognitoTGT** [comment]
I'm gonna downgrade to nextauth stable for the time being, but you can check authjs by reverting to commit [`81aa606`](https://github.com/any-VM/anycode/tree/81aa6061c6331566cd235c4566fe92f31ad26972)

**Comment by chungweileong94** [comment]
It seems like the `authConfig` doesn't have the `callbacks. authorized` option defined, it is used to tell NextAuth how to verify the access.

You can refer to this for further explanation, https://nextjs.org/learn/dashboard-app/adding-authentication#protecting-your-routes-with-nextjs-middleware

**Comment by Lucifer472** [comment]
ok i was having this issue use env file add AUTH_URL with localhost url like this AUTH_URL=http://localhost:3000/api/auth it should fix the issue!

**Comment by rlking1985** [comment]
Similar setup to the OP and having the same issue

**Comment by IncognitoTGT** [comment]
So it turns out that middleware only updates the session if setup like that, I just added an auth callback to solve the issue

```ts
import { config as authConfig } from "@/lib/auth.config";
import NextAuth from "next-auth";
import { NextResponse } from "next/server";
const { auth } = NextAuth(authConfig);
export const middleware = auth(async (req) => {
	if (
		req.auth ||
		req.nextUrl.pathname.startsWith("/auth/login") ||
		req.nextUrl.pathname.startsWith("/auth/error") ||
		req.nextUrl.pathname.startsWith("/auth/verify")
	) {
		return NextResponse.next();
	}
	const url = new URL("/auth/login", req.url);
	url.searchParams.set("callbackUrl", req.nextUrl.pathname);
	return NextResponse.redirect(url);
});
export const config = {
	matcher: ["/((?!_next/static|_next/image|icon.svg|websockify|api/auth|manifest.webmanifest).*)"],
};
```
